src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -6355,6 +6408,416 @@ register_handler(int sig, struct sigaction *old_handler)
+		return EBUSY;
+	}
+
+	D_MUTEX_LOCK(&aio_ctx->lock);
+	if (aio_ctx->inited == false) {
+		D_ALLOC_ARRAY(aio_ctx->ev_done_list, aio_ctx->depth);
+		if (aio_ctx->ev_done_list == NULL) {
+			D_MUTEX_UNLOCK(&aio_ctx->lock);
+			return ENOMEM;
+		}
+		/* empty queue */
+		aio_ctx->ev_done_h       = 0;
+		aio_ctx->ev_done_t       = 0;
+		aio_ctx->ev_done_list[0] = NULL;
+		aio_ctx->ev_queue_len    = 0;
+		aio_ctx->inited          = true;
+	}
+	aio_ctx->on_dfs = true;
+	D_MUTEX_UNLOCK(&aio_ctx->lock);
+
+	return 0;
+
+free_eq:
+	for (j = 0; j < i; j++) {
+		daos_eq_destroy(aio_eq_list[idx + j].eq, 0);
+	}
+	D_MUTEX_UNLOCK(&lock_aio);
+	return daos_der2errno(rc);
+}
+
+int
+io_submit(io_context_t ctx, long nr, struct iocb *ios[])
+{
+	aio_ctx_t     *aio_ctx_obj = (aio_ctx_t *)ctx;
+	io_context_t   ctx_real    = aio_ctx_obj->ctx;
+	daos_size_t    read_size   = 0;
+	d_iov_t        iov         = {};
+	d_sg_list_t    sgl         = {};
+	struct aio_ev *ctx_ev;
+	int            i, n_op_dfs, fd, io_depth;
+	int            idx_aio_eq;
+	int            rc;
+	short          op;
+	long int       aio_eq_count_local;
+
+	if (next_io_submit == NULL) {
+		next_io_submit = dlsym(RTLD_NEXT, "io_submit");
+		D_ASSERT(next_io_submit != NULL);
+	}
+	if (!hook_enabled)
+		return next_io_submit(ctx_real, nr, ios);
+	io_depth = aio_ctx_obj->depth;
+	if (io_depth == 0)
+		return next_io_submit(ctx_real, nr, ios);
+
+	n_op_dfs = 0;
+	for (i = 0; i < nr; i++) {
+		if (ios[i]->aio_fildes >= FD_FILE_BASE)
+			n_op_dfs++;
+	}
+	if (n_op_dfs == 0)
+		return next_io_submit(ctx_real, nr, ios);
+
+	if (n_op_dfs != nr) {
+		DS_ERROR(EINVAL, "io_submit() does not support mixed non-dfs and dfs files yet");
+		return (-EINVAL);
+	}
+
+	/* only support IO_CMD_PREAD and IO_CMD_PWRITE */
+	for (i = 0; i < nr; i++) {
+		op = ios[i]->aio_lio_opcode;
+		if (op != IO_CMD_PREAD && op != IO_CMD_PWRITE) {
+			DS_ERROR(EINVAL, "io_submit only supports PREAD and PWRITE for now");
+			return (-EINVAL);
+		}
+	}
+
+	aio_eq_count_local = aio_eq_count;
+	for (i = 0; i < nr; i++) {
+		op         = ios[i]->aio_lio_opcode;
+		fd         = ios[i]->aio_fildes - FD_FILE_BASE;
+		idx_aio_eq = (aio_ctx_obj->first_eq + i) % aio_eq_count_local;
+
+		D_ALLOC_PTR(ctx_ev);
+		if (ctx_ev == NULL) {
+			if (i == 0)
+				return -ENOMEM;
+			else
+				return i;
+		}
+
+		rc = daos_event_init(&ctx_ev->ev, aio_eq_list[idx_aio_eq].eq, NULL);
+		if (rc) {
+			D_FREE(ctx_ev);
+			DL_ERROR(rc, "daos_event_init() failed");
+			D_GOTO(err, rc = daos_der2errno(rc));
+		}
+		d_iov_set(&iov, (void *)ios[i]->u.c.buf, ios[i]->u.c.nbytes);
+		sgl.sg_nr     = 1;
+		sgl.sg_iovs   = &iov;
+		ctx_ev->piocb = ios[i];
+		/* EQs are shared by contexts. Need to save ctx when polling EQs. */
+		ctx_ev->ctx = aio_ctx_obj;
+		if (op == IO_CMD_PREAD) {
+			rc = dfs_read(file_list[fd]->dfs_mt->dfs, file_list[fd]->file, &sgl,
+				      ios[i]->u.c.offset, &read_size, &ctx_ev->ev);
+			if (rc)
+				D_GOTO(err, rc);
```

Comment:
Are there read requests lost on this error path?  If the I/O has been submitted then the reads could continue after this function but the user could presumably assume the buffer is not in use and free/re-assign it?

Reply:
@ashleypittman I just added "daos_event_fini(&ctx_ev->ev);" at here to destroy the event. Is this you were suggesting? Thank you!